### ISSUE TITLE
[ALEA] Compilation failure with OpenMPI 1.6

### ISSUE THREAD

**Issue by krivenko** [ALPSCore/ALPSCore#471]
OpenMPI 1.6 declares the first argument of `MPI_Reduce` as `void*` while it should normally be `const void*` (contents of the send buffer is not supposed to be changed).
```
In file included from /home/travis/build/krivenko/ALPSCore/alea/test/mpi_twogauss.cpp:1:
/home/travis/build/krivenko/ALPSCore/alea/include/alps/alea/mpi.hpp:90:22: error: 
      no matching function for call to 'MPI_Reduce'
        mpi::checked(MPI_Reduce(sendbuf, data.data(), data.size(), dtype_tag,
                     ^~~~~~~~~~
/home/travis/build/krivenko/ALPSCore/alea/include/alps/alea/mpi.hpp:65:53: note: 
      in instantiation of function template specialization
      'alps::alea::mpi_reducer::inplace_reduce<double>' requested here
    void reduce(view<double> data) const override { inplace_reduce(data); }
                                                    ^
/usr/lib/openmpi/include/mpi.h:1352:20: note: candidate function not viable: 1st
      argument ('const void *') would lose const qualifier
OMPI_DECLSPEC  int MPI_Reduce(void *sendbuf, void *recvbuf, int count,
                   ^
In file included from /home/travis/build/krivenko/ALPSCore/alea/test/mpi_twogauss.cpp:1:
/home/travis/build/krivenko/ALPSCore/alea/include/alps/alea/mpi.hpp:90:22: error: 
      no matching function for call to 'MPI_Reduce'
        mpi::checked(MPI_Reduce(sendbuf, data.data(), data.size(), dtype_tag,
                     ^~~~~~~~~~
/home/travis/build/krivenko/ALPSCore/alea/include/alps/alea/mpi.hpp:67:51: note: 
      in instantiation of function template specialization
      'alps::alea::mpi_reducer::inplace_reduce<long>' requested here
    void reduce(view<long> data) const override { inplace_reduce(data); }
                                                  ^
/usr/lib/openmpi/include/mpi.h:1352:20: note: candidate function not viable: 1st
      argument ('const void *') would lose const qualifier
OMPI_DECLSPEC  int MPI_Reduce(void *sendbuf, void *recvbuf, int count,
                   ^
2 errors generated.
```

**Comment by mwallerb** [comment]
Yeah, that sucks. (Also: 1.6?! Hello, Marty McFly!)

Anybody a good idea shy of ugly preprocessor switches? (@krivenko?)

**Comment by galexv** [comment]
Suggestion 1: Detect the version of OpenMPI (at CMake time), store it in `config.hpp`,  and do `const_cast` if the version is old. Also, hide this inside `alps::mpi`.
Suggestion 2: Refuse to accept OpenMPI 1.6 at CMake phase (because it's a "retired" version) and argue that since May 2012 one could as well update the MPI version. I like this option a little less.

**Comment by krivenko** [comment]
> Also: 1.6?! Hello, Marty McFly!

That's what one gets with `libopenmpi-dev` package on Travis CI `trusty` image.

> Anybody a good idea shy of ugly preprocessor switches? (@krivenko?)

The old accumulators dealt with the problem in a bold way.
```c++
// WORKAROUND:
// for some reason, OpenMPI 1.6 declares `sendbuf` as `void*`, hence `const_cast`.
const int rc=MPI_Reduce(const_cast<void*>(sendbuf), recvbuf, count, datatype, op, root, comm);
```
Emanuel @egull suggest the same approach here.
I don't have a personal preference.

**Comment by iskakoff** [comment]
@mwallerb Since ALPSCore should be compatible with MPI-2.0 standard and `const` modifier for send buffers was introduced only since MPI-3.0, we need to have `const_cast`.

**Comment by iskakoff** [comment]
fixed by pr #474
